nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSCachingRepository.java

```diff
@@ -0,0 +1,399 @@
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.PartialList;
+import org.nuxeo.ecm.core.api.ScrollResult;
+import org.nuxeo.ecm.core.blob.BlobManager;
+import org.nuxeo.ecm.core.model.LockManager;
+import org.nuxeo.ecm.core.model.Session;
+import org.nuxeo.ecm.core.query.sql.model.OrderByClause;
+import org.nuxeo.ecm.core.storage.FulltextConfiguration;
+import org.nuxeo.ecm.core.storage.State;
+import org.nuxeo.ecm.core.storage.State.StateDiff;
+import org.nuxeo.runtime.metrics.MetricsService;
+
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.SharedMetricRegistries;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Ordering;
+
+/**
+ * The DBS Cache layer used to cache some method call of real repository
+ *
+ * @since 8.10
+ */
+public class DBSCachingRepository implements DBSRepository {
+
+    private static final Log log = LogFactory.getLog(DBSCachingRepository.class);
+
+    private static final Random RANDOM = new Random();
+
+    private final DBSRepository repository;
+
+    private final Cache<String, State> cache;
+
+    private final Cache<String, String> childCache;
+
+    private DBSClusterInvalidator clusterInvalidator;
+
+    private final DBSInvalidations invalidations;
+
+    protected final MetricRegistry registry = SharedMetricRegistries.getOrCreate(MetricsService.class.getName());
+
+    public DBSCachingRepository(DBSRepository repository, DBSRepositoryDescriptor descriptor) {
+        this.repository = repository;
+        // Init caches
+        cache = newCache(descriptor);
+        registry.registerAll(GuavaCacheMetric.of(cache, "nuxeo", "repositories", repository.getName(), "cache"));
+        childCache = newCache(descriptor);
+        registry.registerAll(
+                GuavaCacheMetric.of(childCache, "nuxeo", "repositories", repository.getName(), "childCache"));
+        invalidations = new DBSInvalidations();
+        if (descriptor.isClusteringEnabled()) {
+            initClusterInvalidator(descriptor);
+        }
+    }
+
+    protected <T> Cache<String, T> newCache(DBSRepositoryDescriptor descriptor) {
+        CacheBuilder<Object, Object> builder = CacheBuilder.newBuilder();
+        builder = builder.expireAfterWrite(descriptor.cacheTtl, TimeUnit.MINUTES);
+        if (descriptor.cacheConcurrencyLevel != null) {
+            builder = builder.concurrencyLevel(descriptor.cacheConcurrencyLevel);
+        }
+        if (descriptor.cacheMaxSize != null) {
+            builder = builder.maximumSize(descriptor.cacheMaxSize);
+        }
+        return builder.build();
+    }
+
+    protected void initClusterInvalidator(DBSRepositoryDescriptor descriptor) {
+        String nodeId = descriptor.clusterNodeId;
+        if (StringUtils.isBlank(nodeId)) {
+            nodeId = String.valueOf(RANDOM.nextInt(32768));
```

Comment:
Please use a bigger max value. We were limited to small integers for VCS for historical reasons, but there's no such limit here.